rfd/0095-okta-integration.md

```diff
@@ -0,0 +1,405 @@
+      # Can be any user ID supported by Okta, documented in
+      # https://developer.okta.com/docs/reference/api/users/#get-user
+      okta_user_id: alice@mydomain.com
+```
+
+#### Launching Okta applications
+
+[Application objects from Okta](https://developer.okta.com/docs/reference/api/apps/#application-object) can return one or more `appLinks` which, when launched directly from
+Teleport, will forward to the expected application along with Okta authentication. These links
+will be used as the application URLs synced from Okta.
+
+**Note**: These links will be redirected to their actual external URLs rather than living behind
+Teleport's proxy.
+
+### Backend modifications
+
+A few modifications to existing services, databases, and models will be necessary to support this
+functinoality.
+
+#### Applications
+
+Applications will need some new information to support this:
+
+* A new flag to indicate that the application has been externally synchronized. This will allow
+  Teleport to distinguish between regular Teleport applications and external applications.
+* User specific application metadata for external applications. This will dictate to the UI what
+  kind of options to display and whether the URLs are accessible. These will be joined with the
+  application objects during query. To start, the only necessary fields for this object are a
+  user mapping, a corresponding external identity provider name, and a boolean flag as to
+  whether the user has access.
+
+#### External groups
+
+A new notion of external groups will need to be added. This represent external groups known to
+an identity provider. Additionally, similar to applications, there must be a notion of user
+specific group metadata that contains information as to whether a user is a member of the group.
+
+#### Users
+
+Users will need new provider specific configuration fields. These will necessarily be pretty
+free-form.
+
+#### Request access service
+
+The requests for external identity providers will be longer term than Teleport's current notion
+of requests. The service will either need to be expanded to support long term requests or
+turned into a separate service that supports access requests that cannot be later dropped.
+
+### Security
+
+* It's possible to capture sensitive data in our internal Teleport database. This could be
+  undesirable for customers or other users of Teleport. We should take care that the sensitive
+  identity provider data (like Okta API tokens) are never communicated to or stored outside of
+  the configuration file.
+* The mapping of Okta users to the appropriate Teleport user is a concern. If a user has been
+  mismapped to an Okta user, it will be possible for that user to request access on behalf of
+  that user. Administrators should take care with their Okta identity mappings, or even better,
+  ensure that Teleport logins map directly to Okta users.
+
+### Open questions
+
+#### Is the identity provider abstraction overkill?
```

Comment:
I thought about the same. I don't know if we'll want to expand. FWIW I think defining "IdP" interface in the code like mentioned somewhere above is ok. Might be something we can chat about with the product team.